src/Azure/Orleans.Transactions.AzureStorage/TransactionalState/AzureTableTransactionalStateStorage.cs

```diff
@@ -232,106 +243,128 @@ private bool FindState(long sequenceId, out int pos)
+            await foreach (var result in queryResult)
+            {
+                return result;
+            }
+
+            return new KeyEntity()
+            {
+                PartitionKey = partition,
+                RowKey = KeyEntity.RK
+            };
         }
 
         private async Task<List<KeyValuePair<long, StateEntity>>> ReadStates()
         {
-            var query = new TableQuery<StateEntity>()
-                .Where(AzureTableUtils.RangeQuery(this.partition, StateEntity.RK_MIN, StateEntity.RK_MAX));
-            TableContinuationToken continuationToken = null;
+            var query = AzureTableUtils.RangeQuery(this.partition, StateEntity.RK_MIN, StateEntity.RK_MAX);
             var results = new List<KeyValuePair<long, StateEntity>>();
-            do
+            var queryResult = table.QueryAsync<TableEntity>(query).ConfigureAwait(false);
+            await foreach (var entity in queryResult)
             {
-                TableQuerySegment<StateEntity> queryResult = await table.ExecuteQuerySegmentedAsync(query, continuationToken).ConfigureAwait(false);
-                foreach (var x in queryResult.Results)
-                {
-                    results.Add(new KeyValuePair<long, StateEntity>(x.SequenceId, x));
-                };
-                continuationToken = queryResult.ContinuationToken;
-            } while (continuationToken != null);
+                var state = new StateEntity(entity);
+                results.Add(new KeyValuePair<long, StateEntity>(state.SequenceId, state));
+            };
             return results;
         }
 
         private class BatchOperation
         {
-            private readonly TableBatchOperation batchOperation;
+            private readonly List<TableTransactionAction> batchOperation;
             private readonly ILogger logger;
-            private readonly KeyEntity key;
-            private readonly CloudTable table;
+            private readonly TableClient table;
+            private KeyEntity key;
 
-            private bool batchContainsKey;
+            private int keyIndex = -1;
 
-            public BatchOperation(ILogger logger, KeyEntity key, CloudTable table)
+            public BatchOperation(ILogger logger, KeyEntity key, TableClient table)
             {
-                this.batchOperation = new TableBatchOperation();
+                this.batchOperation = new();
                 this.logger = logger;
                 this.key = key;
                 this.table = table;
             }
 
-            public async Task Add(TableOperation operation)
-            {
-                batchOperation.Add(operation);
+            public ETag KeyETag => key.ETag;
+            private bool BatchHasKey => keyIndex >= 0;
 
-                if (operation.Entity == key)
+            public async Task Add(TableTransactionAction operation)
```

Comment:
suggestion: Make BatchOperation.Add return `ValueTask` since it will mostly return synchronously (assuming the need to call Flush is the less common case)